aws_lambda_powertools/event_handler/api_gateway.py

```diff
@@ -218,13 +221,32 @@ def __init__(
         cors: bool,
         compress: bool,
         cache_control: Optional[str],
+        middleware: Optional[List[Callable]],
     ):
         self.method = method.upper()
         self.rule = rule
         self.func = func
         self.cors = cors
         self.compress = compress
         self.cache_control = cache_control
+        self.middleware = middleware or []
+
+    def __call__(self, router_middleware: List[Callable], app, args: Dict[str, str]):
+        """Builds the middleware stack using global and route middlewares, redefining the original handler function"""
+        if not self._middleware_built:
+            # prepend global router middleware first
+            all_middleware = list(router_middleware) + list(self.middleware)
+
+            # IMPORTANT: # this must be the last mdidleware in the stack to avoid breaking changes
+            # for the registered API call signature (Maintain Backward Compatibility)
+            all_middleware.append(registered_api_middleware)
+
+            for handler in reversed(all_middleware):
+                self.func = MiddlewareHandler(handler=handler, next_handler=self.func)
+
+            self._middleware_built = True
+
+        return self.func(app, **args)
```

Comment:
We're missing a short-circuit mechanism so a middleware could return early - e.g., raise a HTTP error of sorts.

Reply:
don't need an explicit short-circuit mechanism - any middleware handler raising an exception would cause the Python call stack to unwind naturally - and any existing `try...except` handling in middleware below the exception would execute as expected.  I feel this is where really detailed documentation is required which is needed - I wanted to get this out for discussion and collaboration.

Reply:
Is the confusion the code loop?  the loop is actually "wrapping" the original function handler using a chain of `MiddlewareHandler` class wrappers in reverse order (inside-out), which replaces the actual instance of `self.func` with the outermost version of MiddlewareHandler class which is executed at line 249 via the python `__call__` class implementation - this is very much in-line with Chalice and Starlette middleware handling (I did lots of reading last week).